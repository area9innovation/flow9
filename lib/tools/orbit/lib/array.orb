// Array utility functions library for Orbit

// map: Apply a function to each element in an array
fn map(arr, f) = (
    if length(arr) = 0 then []
    else [f(arr[0])] + map(subrange(arr, 1, length(arr) - 1), f)
);

// filter: Keep only elements that match a predicate
fn filter(arr, pred) = (
    if length(arr) = 0 then []
    else if pred(arr[0]) then 
        [arr[0]] + filter(subrange(arr, 1, length(arr) - 1), pred)
    else 
        filter(subrange(arr, 1, length(arr) - 1), pred)
);

// fold: Accumulate a result based on each array element and an accumulator
fn fold(arr, init, f) = (
    if length(arr) = 0 then init
    else fold(subrange(arr, 1, length(arr) - 1), f(init, arr[0]), f)
);

// mapi: Map with index
fn mapi(arr, f) = (
    fn mapiHelper(arr, f, idx) = (
        if length(arr) = 0 then []
        else [f(arr[0], idx)] + mapiHelper(subrange(arr, 1, length(arr) - 1), f, idx + 1)
    );
    mapiHelper(arr, f, 0)
);

// filteri: Filter with index
fn filteri(arr, pred) = (
    fn filteriHelper(arr, pred, idx) = (
        if length(arr) = 0 then []
        else if pred(arr[0], idx) then 
            [arr[0]] + filteriHelper(subrange(arr, 1, length(arr) - 1), pred, idx + 1)
        else 
            filteriHelper(subrange(arr, 1, length(arr) - 1), pred, idx + 1)
    );
    filteriHelper(arr, pred, 0)
);

// foldi: Fold with index
fn foldi(arr, init, f) = (
    fn foldiHelper(arr, acc, f, idx) = (
        if length(arr) = 0 then acc
        else foldiHelper(subrange(arr, 1, length(arr) - 1), f(acc, arr[0], idx), f, idx + 1)
    );
    foldiHelper(arr, init, f, 0)
);

// tail: Return all elements except the first one
fn tail(arr) = (
    if length(arr) <= 1 then []
    else subrange(arr, 1, length(arr) - 1)
);

// tailFrom: Return all elements from a given index to the end
fn tailFrom(arr, idx) = (
    if idx >= length(arr) then []
    else subrange(arr, idx, length(arr) - idx)
);

// take: Return the first n elements of an array
fn take(arr, n) = (
    if n <= 0 || length(arr) = 0 then []
    else [arr[0]] + take(subrange(arr, 1, length(arr) - 1), n - 1)
);

// iter: Call a function for each element (for side effects)
fn iter(arr, f) = (
    if length(arr) = 0 then 0 // Return 0 instead of ()
    else (
        f(arr[0]);
        iter(subrange(arr, 1, length(arr) - 1), f)
    )
);

// iteri: Iterate with index
fn iteri(arr, f) = (
    fn iteriHelper(arr, f, idx) = (
        if length(arr) = 0 then 0 // Return 0 instead of ()
        else (
            f(arr[0], idx);
            iteriHelper(subrange(arr, 1, length(arr) - 1), f, idx + 1)
        )
    );
    iteriHelper(arr, f, 0)
);

// filtermap: Combine filter and map operations
fn filtermap(arr, f) = (
    if length(arr) = 0 then []
    else (
        let result = f(arr[0]);
        result is (
            Some(val) => [val] + filtermap(subrange(arr, 1, length(arr) - 1), f);
            None() => filtermap(subrange(arr, 1, length(arr) - 1), f);
            _ => filtermap(subrange(arr, 1, length(arr) - 1), f) // For safety
        )
    )
);

// filtermapi: Filtermap with index
fn filtermapi(arr, f) = (
    fn filtermapHelper(arr, f, idx) = (
        if length(arr) = 0 then []
        else (
            let result = f(arr[0], idx);
            result is (
                Some(val) => [val] + filtermapHelper(subrange(arr, 1, length(arr) - 1), f, idx + 1);
                None() => filtermapHelper(subrange(arr, 1, length(arr) - 1), f, idx + 1);
                _ => filtermapHelper(subrange(arr, 1, length(arr) - 1), f, idx + 1) // For safety
            )
        )
    );
    filtermapHelper(arr, f, 0)
);

// removeIndex: Remove an element at a specific index
fn removeIndex(arr, idx) = (
    if idx < 0 || idx >= length(arr) then arr
    else take(arr, idx) + tailFrom(arr, idx + 1)
);

// removeRange: Remove a range of elements
fn removeRange(arr, start, count) = (
    if start < 0 || start >= length(arr) || count <= 0 then arr
    else take(arr, start) + tailFrom(arr, start + count)
);

// removeFirst: Remove the first occurrence of a specific element
fn removeFirst(arr, element) = (
    fn removeFirstHelper(before, after) = (
        if length(after) = 0 then before
        else if after[0] = element then before + subrange(after, 1, length(after) - 1)
        else removeFirstHelper(before + [after[0]], subrange(after, 1, length(after) - 1))
    );
    removeFirstHelper([], arr)
);

// removeAll: Remove all occurrences of a specific element
fn removeAll(arr, element) = (
    filter(arr, \x.x != element != 0)
);

// contains: Check if an element exists in the array
fn contains(arr, element) = (
    if length(arr) = 0 then false
    else if arr[0] = element then true
    else contains(subrange(arr, 1, length(arr) - 1), element)
);

// exists: Check if any element satisfies a predicate
fn exists(arr, pred) = (
    if length(arr) = 0 then false
    else if pred(arr[0]) then true
    else exists(subrange(arr, 1, length(arr) - 1), pred)
);

// forall: Check if all elements satisfy a predicate
fn forall(arr, pred) = (
    if length(arr) = 0 then true
    else if !pred(arr[0]) then false
    else forall(subrange(arr, 1, length(arr) - 1), pred)
);

// uniq: Remove duplicates from an array
fn uniq(arr) = (
    if length(arr) <= 1 then arr
    else (
        fn uniq_helper(remaining, result) = (
            if length(remaining) = 0 then result
            else (
                let current = remaining[0];
                let rest = tail(remaining);
                
                if contains(result, current) then
                    // Skip duplicate
                    uniq_helper(rest, result)
                else
                    // Keep this element
                    uniq_helper(rest, result + [current])
            )
        );
        uniq_helper(arr, [])
    )
);