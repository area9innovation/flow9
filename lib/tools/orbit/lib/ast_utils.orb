// AST utility functions for Orbit

// Create a properly nested AST node by folding over an array of arguments
// This function handles arrays of any length, whereas makeAst expects exactly 2 arguments
fn makeFoldedAst(op, args) = (
    let len = length(args);
    
    if len = 0 then (
        // Handle empty array case - return an appropriate default value
        // This depends on the operator, but we'll return a neutral element
        if op = "&&" then true
        else if op = "||" then false
        else args  // Return empty array as fallback
    ) else if len = 1 then (
        // For single element, just return it without creating an operation
        args[0]
    ) else if len = 2 then (
        // For exactly two elements, we can use makeAst directly
        makeAst(op, args)
    ) else (
        // For 3+ elements, fold to create a proper binary tree
        // Start with the first element and fold over the rest
        let initialAst = args[0];
        let remainingArgs = subrange(args, 1, len - 1);
        
        // Fold over the remaining arguments, combining each with the accumulator
        fold(remainingArgs, initialAst, \acc, element -> (
            makeAst(op, [acc, element])
        ))
    )
);