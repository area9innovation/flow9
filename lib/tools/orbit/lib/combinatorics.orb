// Basic combinatorial functions
fn factorial(n) = (
	if n <= 1 then 1
	else n * factorial(n - 1)
);

fn binomial_coefficient(n, k) = (
	if k < 0 || k > n then 0
	else if k == 0 || k == n then 1
	else (
		// Use multiplicative formula to avoid large intermediate values
		fn binom_helper(i, result) = (
			if i > k then result
			else binom_helper(i + 1, result * (n - k + i) / i)
		);
		binom_helper(1, 1)
	)
);

fn permutations(n, r) = (
	if r < 0 || r > n then 0
	else if r == 0 then 1
	else (
		// Calculate n! / (n-r)!
		fold(range(n-r+1, n+1), 1, \product, i.product * i)
	)
);

// Generate permutations of an array
fn generate_permutations(arr) = (
	if length(arr) <= 1 then [arr]
	else (
		let result = [];
		fn perms_helper(i, perms) = (
			if i >= length(arr) then perms
			else (
				let element = arr[i];
				let rest = remove_at_index(arr, i);
				let sub_perms = generate_permutations(rest);

				let new_perms = fold(sub_perms, perms, \acc, p.acc + [[element] + p]);
				perms_helper(i + 1, new_perms)
			)
		);
		perms_helper(0, [])
	)
);

// Generate combinations of an array
fn generate_combinations(arr, r) = (
	if r > length(arr) then []
	else if r == 0 then [[]]
	else if r == length(arr) then [arr]
	else (
		if length(arr) == 0 then []
		else (
			let first = arr[0];
			let rest = subrange(arr, 1, length(arr) - 1);

			// Combinations that include the first element
			let with_first = map(generate_combinations(rest, r - 1),
													\comb.[first] + comb);

			// Combinations that exclude the first element
			let without_first = generate_combinations(rest, r);

			with_first + without_first
		)
	)
);
