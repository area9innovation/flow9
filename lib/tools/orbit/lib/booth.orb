import lib/array;

// Linear time Booth's algorithm to find canonical form in cyclic groups
fn booth_canonical(arr) = (
	if length(arr) <= 1 then arr
	else (
		let n = length(arr);

		// Create an auxiliary array which is the original array concatenated with itself
		let double_arr = arr + arr;

		// Initialize failure function array with -1
		// (We use -1 instead of conventional 0 to simplify logic)
		let f = \i. if i < 0 || i >= n then -1 else f(i);

		// Compute the min rotation index in linear time
		fn compute_min_index() = (
			// Initialize starting values
			fn find_min(i, j, k) = (
				// We're done when we've checked all rotations
				if i + k >= 2 * n then j
				else if k >= n then find_min(i + 1, j, 0)
				else (
					// Compute the current indices we're comparing
					let i_idx = (i + k) % (2 * n);
					let j_idx = (j + k) % (2 * n);

					if double_arr[i_idx] < double_arr[j_idx] then
						// i rotation is smaller, j becomes i
						find_min(i, i, k + 1)
					else if double_arr[i_idx] > double_arr[j_idx] then
						// j rotation is smaller, i moves forward
						find_min(i + k + 1, j, 0)
					else
						// Equal so far, continue comparison
						find_min(i, j, k + 1)
				)
			);

			// Start with first two positions
			find_min(0, 0, 0)
		);

		let min_start = compute_min_index();

		// Build the canonical form starting from min_start
		fn build_result(idx, result) = (
			if idx >= n then result
			else build_result(idx + 1, result + [arr[(min_start + idx) % n]])
		);

		build_result(0, [])
	)
);

