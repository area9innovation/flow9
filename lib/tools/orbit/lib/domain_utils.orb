import lib/graph;

// Compare two domains in the hierarchy
// Returns: 
//   -1 if a is more specific than b (a ⊂ ... ⊂ b)
//    0 if a and b are the same
//    1 if b is more specific than a (b ⊂ ... ⊂ a)
//    None() if a and b are not directly related
fn compareDomains(graph, a, b) = (
    if (a == b) then 0  // Same domain
    else if (isReachable(graph, a, b)) then -1  // a is more specific than b
    else if (isReachable(graph, b, a)) then 1   // b is more specific than a
    else None()  // Not directly related
);

// Find the lowest domain from a list of domains
// The lowest domain is the most specific one in the hierarchy
fn findLowestDomain(graph, domains) = (
    if (length(domains) == 0) then None()
    else if (length(domains) == 1) then Some(domains[0])
    else (
        // Start with the first domain as our current lowest
        let initialLowest = domains[0];
        
        // Compare with each other domain to find the lowest
        fold(subrange(domains, 1, length(domains) - 1), initialLowest, \currentLowest, domain -> (
            compareDomains(graph, domain, currentLowest) is (
                -1 => domain;  // This domain is more specific than current lowest
                0 => currentLowest;  // Same domain, keep current
                1 => currentLowest;  // Current lowest is more specific
                None() => (
                    // Domains are not directly related, we keep the current lowest
                    // A more sophisticated approach might look at the full hierarchy
                    currentLowest
                )
            )
        ))
    )
);

// Get the most specific domain a node belongs to
// This is the main function to be used in the OGraph system
fn getNodeMostSpecificDomain(domainGraph, og, nodeId) = (
    // Get all domains the node belongs to
    let domains = getONodeBelongsTo(og, nodeId);
    
    // Find and return the lowest domain from the list
    findLowestDomain(domainGraph, domains)
);

"Domain Hierarchy Utility Library Successfully Loaded"