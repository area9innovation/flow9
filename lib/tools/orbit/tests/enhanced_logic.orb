import lib/logic;
import lib/bdd;  // For comparison with BDD-based conversion

// Test the enhanced logic canonicalization
fn main() = (
    println("==== Testing Enhanced Logic Canonicalization ====");
    
    // Test cases for DNF conversion
    let dnf_test_cases = [
        // Original example
        ["(a ∨ b) ∧ (¬a ∨ c)", quote((a || b) && (!a || c))],
        
        // Basic simplifications
        ["a ∧ (a ∨ b)", quote(a && (a || b))],
        ["(a ∧ b) ∨ (a ∧ c)", quote((a && b) || (a && c))],
        ["a ∨ (a ∧ b)", quote(a || (a && b))],
        
        // Nested expressions
        ["(a ∨ (b ∨ c)) ∧ d", quote((a || (b || c)) && d)],
        ["a ∧ (b ∧ (c ∧ d))", quote(a && (b && (c && d)))],
        
        // Contradictions and tautologies
        ["(a ∧ ¬a) ∨ b", quote((a && !a) || b)],
        ["a ∧ (b ∨ ¬b)", quote(a && (b || !b))],
        
        // Complex cases
        ["(a ∨ b) ∧ (a ∨ c) ∧ (¬a ∨ d)", quote((a || b) && (a || c) && (!a || d))],
        ["(a ∨ b) ∧ (a ∨ ¬b) ∧ (a ∨ c)", quote((a || b) && (a || !b) && (a || c))],
        ["(a ∧ b) ∨ (¬a ∧ b) ∨ (a ∧ ¬b)", quote((a && b) || (!a && b) || (a && !b))]
    ];
    
    // Test cases for CNF conversion
    let cnf_test_cases = [
        // Special CNF pattern
        ["(a ∧ b) ∨ (¬a ∧ c)", quote((a && b) || (!a && c))],
        
        // Basic CNF simplifications
        ["a ∨ (a ∧ b)", quote(a || (a && b))],
        ["(a ∨ b) ∧ (a ∨ c)", quote((a || b) && (a || c))],
        
        // Complex CNF cases
        ["(a ∧ b) ∨ c", quote((a && b) || c)],
        ["a ∨ (b ∧ c)", quote(a || (b && c))],
        ["(a ∨ b) ∧ (a ∨ ¬b) ∧ (a ∨ c)", quote((a || b) && (a || !b) && (a || c))],
        ["(a ∧ b) ∨ (¬a ∧ c) ∨ (a ∧ d)", quote((a && b) || (!a && c) || (a && d))]
    ];
    
    // Test DNF conversion
    println("\n----- DNF Conversion Tests -----");
    fold(dnf_test_cases, 1, \(i, test) -> (
        let name = test[0];
        let expr = test[1];
        println("\nTest " + i2s(i) + ": " + name);
        println("Original: " + prettyOrbit(expr));
		let dnf = to_dnf(expr);
        println("DNF: " + prettyOrbit(dnf));
        
        // Compare with BDD-based conversion for validation
        let bdd_result = from_bdd(to_bdd(expr));
        println("BDD DNF: " + prettyOrbit(bdd_result));
        
        // Check if they're equivalent
        let equivalent = are_equivalent(dnf, expr);
        println("Equivalent to original: " + (if equivalent then "Yes" else "No"));
        
        i + 1
    ));
    
    // Test CNF conversion
    println("\n----- CNF Conversion Tests -----");
    fold(cnf_test_cases, 1, \(i, test) -> (
        let name = test[0];
        let expr = test[1];
        println("\nTest " + i2s(i) + ": " + name);
        println("Original: " + prettyOrbit(expr));
        println("CNF: " + prettyOrbit(to_cnf(expr)));
        
        // Compare with simplified BDD (not true CNF but good for equivalence)
        let bdd_result = from_bdd(to_bdd(expr));
        println("BDD simplified: " + prettyOrbit(bdd_result));
        
        // Check if they're equivalent
        let equivalent = are_equivalent(to_cnf(expr), expr);
        println("Equivalent to original: " + (if equivalent then "Yes" else "No"));
        
        i + 1
    ));
    
    "Done"
);

// Run the tests
main();