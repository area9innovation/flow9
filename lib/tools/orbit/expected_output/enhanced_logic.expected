==== Testing Enhanced Logic Canonicalization ====
----- DNF Conversion Tests -----
Test 1: (a ∨ b) ∧ (¬a ∨ c)
Original: ((a ∨ b) ∧ (¬a ∨ c))
Rule 'special_dnf_case' applied. ID: 12: ((a ∧ c) ∨ (¬a ∧ b)) was ((a ∨ b) ∧ (¬a ∨ c)) at 8
DNF: ((a ∧ c) ∨ (¬a ∧ b))
BDD DNF: ((a ∧ c) ∨ (¬a ∧ b))
Equivalent to original: Yes
Test 2: a ∧ (a ∨ b)
Original: (a ∧ (a ∨ b))
Rule 'absorption3' applied. ID: 1: a was (a ∧ (a ∨ b)) at 5
DNF: a
BDD DNF: a
Equivalent to original: Yes
Test 3: (a ∧ b) ∨ (a ∧ c)
Original: ((a ∧ b) ∨ (a ∧ c))
DNF: ((a ∧ b) ∨ (a ∧ c))
BDD DNF: (a ∧ (b ∨ c))
Equivalent to original: Yes
Test 4: a ∨ (a ∧ b)
Original: (a ∨ (a ∧ b))
Rule 'absorption1' applied. ID: 1: a was (a ∨ (a ∧ b)) at 5
DNF: a
BDD DNF: a
Equivalent to original: Yes
Test 5: (a ∨ (b ∨ c)) ∧ d
Original: ((a ∨ (b ∨ c)) ∧ d)
Rule 'dist_and_over_or2' applied. ID: 10: ((a ∧ d) ∨ ((b ∨ c) ∧ d)) was ((a ∨ (b ∨ c)) ∧ d) at 7
Rule 'dist_and_over_or2' applied. ID: 13: ((b ∧ d) ∨ (c ∧ d)) was ((b ∨ c) ∧ d) at 9
DNF: (((a ∧ d) ∨ (b ∧ d)) ∨ (c ∧ d))
BDD DNF: ((a ∧ d) ∨ (¬a ∧ ((b ∧ d) ∨ (¬b ∧ (c ∧ d)))))
Equivalent to original: Yes
Test 6: a ∧ (b ∧ (c ∧ d))
Original: (a ∧ (b ∧ (c ∧ d)))
DNF: (a ∧ (b ∧ (c ∧ d)))
BDD DNF: (a ∧ (b ∧ (c ∧ d)))
Equivalent to original: Yes
Test 7: (a ∧ ¬a) ∨ b
Original: ((a ∧ ¬a) ∨ b)
Rule 'contradiction' applied. ID: 7: false was (a ∧ ¬a) at 4
Rule 'or_with_false' applied. ID: 5: b was (false ∨ b) at 6
DNF: b
BDD DNF: b
Equivalent to original: Yes
Test 8: a ∧ (b ∨ ¬b)
Original: (a ∧ (b ∨ ¬b))
Rule 'tautology' applied. ID: 7: true was (b ∨ ¬b) at 5
Rule 'and_with_true2' applied. ID: 1: a was (a ∧ true) at 6
DNF: a
BDD DNF: a
Equivalent to original: Yes
Test 9: (a ∨ b) ∧ (a ∨ c) ∧ (¬a ∨ d)
Original: (((a ∨ b) ∧ (a ∨ c)) ∧ (¬a ∨ d))
Rule 'advanced_simplify4' applied. ID: 17: ((a ∧ d) ∨ (¬a ∧ (b ∧ c))) was (((a ∨ b) ∧ (a ∨ c)) ∧ (¬a ∨ d)) at 12
Rule 'dist_and_over_or' applied. ID: 20: (((a ∨ b) ∧ a) ∨ ((a ∨ b) ∧ c)) was ((a ∨ b) ∧ (a ∨ c)) at 7
Rule 'dist_and_over_or2' applied. ID: 23: ((a ∧ a) ∨ (b ∧ a)) was ((a ∨ b) ∧ a) at 18
Rule 'dist_and_over_or2' applied. ID: 26: ((a ∧ c) ∨ (b ∧ c)) was ((a ∨ b) ∧ c) at 19
Rule 'and_idempotent' applied. ID: 1: a was (a ∧ a) at 21
DNF: ((a ∧ d) ∨ (¬a ∧ (b ∧ c)))
BDD DNF: ((a ∧ d) ∨ (¬a ∧ (b ∧ c)))
Equivalent to original: Yes
Test 10: (a ∨ b) ∧ (a ∨ ¬b) ∧ (a ∨ c)
Original: (((a ∨ b) ∧ (a ∨ ¬b)) ∧ (a ∨ c))
Rule 'exclusive_or_simplify2' applied. ID: 1: a was ((a ∨ b) ∧ (a ∨ ¬b)) at 8
Rule 'dist_and_over_or' applied. ID: 15: ((a ∧ a) ∨ (a ∧ c)) was (a ∧ (a ∨ c)) at 12
Rule 'and_idempotent' applied. ID: 1: a was (a ∧ a) at 13
Rule 'absorption1' applied. ID: 1: a was (a ∨ (a ∧ c)) at 15
DNF: a
BDD DNF: a
Equivalent to original: Yes
Test 11: (a ∧ b) ∨ (¬a ∧ b) ∨ (a ∧ ¬b)
Original: (((a ∧ b) ∨ (¬a ∧ b)) ∨ (a ∧ ¬b))
Rule 'advanced_simplify2' applied. ID: 2: b was ((a ∧ b) ∨ (¬a ∧ b)) at 8
DNF: (b ∨ (a ∧ ¬b))
BDD DNF: (a ∨ b)
Equivalent to original: Yes
----- CNF Conversion Tests -----
Test 1: (a ∧ b) ∨ (¬a ∧ c)
Original: ((a ∧ b) ∨ (¬a ∧ c))
Rule 'special_cnf_case' applied. ID: 17: ((((a ∨ ¬a) ∧ (a ∨ c)) ∧ (b ∨ ¬a)) ∧ (b ∨ c)) was ((a ∧ b) ∨ (¬a ∧ c)) at 8
Rule 'cnf_tautology_elimination' applied. ID: 11: (a ∨ c) was ((a ∨ ¬a) ∧ (a ∨ c)) at 12
Rule 'tautology' applied. ID: 18: true was (a ∨ ¬a) at 10
CNF: (((a ∨ c) ∧ (b ∨ c)) ∧ (b ∨ ¬a))
BDD simplified: ((a ∧ b) ∨ (¬a ∧ c))
Rule 'special_cnf_case' applied. ID: 17: ((((a ∨ ¬a) ∧ (a ∨ c)) ∧ (b ∨ ¬a)) ∧ (b ∨ c)) was ((a ∧ b) ∨ (¬a ∧ c)) at 8
Rule 'cnf_tautology_elimination' applied. ID: 11: (a ∨ c) was ((a ∨ ¬a) ∧ (a ∨ c)) at 12
Rule 'tautology' applied. ID: 18: true was (a ∨ ¬a) at 10
Equivalent to original: Yes
Test 2: a ∨ (a ∧ b)
Original: (a ∨ (a ∧ b))
Rule 'absorption1' applied. ID: 1: a was (a ∨ (a ∧ b)) at 5
CNF: a
BDD simplified: a
Rule 'absorption1' applied. ID: 1: a was (a ∨ (a ∧ b)) at 5
Equivalent to original: Yes
Test 3: (a ∨ b) ∧ (a ∨ c)
Original: ((a ∨ b) ∧ (a ∨ c))
CNF: ((a ∨ b) ∧ (a ∨ c))
BDD simplified: (a ∨ (b ∧ c))
Equivalent to original: Yes
Test 4: (a ∧ b) ∨ c
Original: ((a ∧ b) ∨ c)
Rule 'dist_or_over_and2' applied. ID: 8: ((a ∨ c) ∧ (b ∨ c)) was ((a ∧ b) ∨ c) at 5
CNF: ((a ∨ c) ∧ (b ∨ c))
BDD simplified: ((a ∧ (b ∨ c)) ∨ (¬a ∧ c))
Rule 'dist_or_over_and2' applied. ID: 8: ((a ∨ c) ∧ (b ∨ c)) was ((a ∧ b) ∨ c) at 5
Equivalent to original: Yes
Test 5: a ∨ (b ∧ c)
Original: (a ∨ (b ∧ c))
Rule 'dist_or_over_and' applied. ID: 8: ((a ∨ b) ∧ (a ∨ c)) was (a ∨ (b ∧ c)) at 5
CNF: ((a ∨ b) ∧ (a ∨ c))
BDD simplified: (a ∨ (b ∧ c))
Rule 'dist_or_over_and' applied. ID: 8: ((a ∨ b) ∧ (a ∨ c)) was (a ∨ (b ∧ c)) at 5
Equivalent to original: Yes
Test 6: (a ∨ b) ∧ (a ∨ ¬b) ∧ (a ∨ c)
Original: (((a ∨ b) ∧ (a ∨ ¬b)) ∧ (a ∨ c))
Rule 'exclusive_or_simplify2' applied. ID: 1: a was ((a ∨ b) ∧ (a ∨ ¬b)) at 8
Rule 'absorption3' applied. ID: 1: a was (a ∧ (a ∨ c)) at 12
CNF: a
BDD simplified: a
Rule 'exclusive_or_simplify2' applied. ID: 1: a was ((a ∨ b) ∧ (a ∨ ¬b)) at 8
Rule 'absorption3' applied. ID: 1: a was (a ∧ (a ∨ c)) at 12
Equivalent to original: Yes
Test 7: (a ∧ b) ∨ (¬a ∧ c) ∨ (a ∧ d)
Original: (((a ∧ b) ∨ (¬a ∧ c)) ∨ (a ∧ d))
Rule 'special_cnf_case' applied. ID: 21: ((((a ∨ ¬a) ∧ (a ∨ c)) ∧ (b ∨ ¬a)) ∧ (b ∨ c)) was ((a ∧ b) ∨ (¬a ∧ c)) at 8
Rule 'cnf_tautology_elimination' applied. ID: 15: (a ∨ c) was ((a ∨ ¬a) ∧ (a ∨ c)) at 16
Rule 'dist_or_over_and' applied. ID: 24: (((((a ∨ c) ∧ (b ∨ ¬a)) ∧ (b ∨ c)) ∨ a) ∧ ((((a ∨ c) ∧ (b ∨ ¬a)) ∧ (b ∨ c)) ∨ d)) was ((((a ∨ c) ∧ (b ∨ ¬a)) ∧ (b ∨ c)) ∨ (a ∧ d)) at 12
Rule 'dist_or_over_and2' applied. ID: 27: ((((a ∨ c) ∧ (b ∨ ¬a)) ∨ a) ∧ ((b ∨ c) ∨ a)) was ((((a ∨ c) ∧ (b ∨ ¬a)) ∧ (b ∨ c)) ∨ a) at 22
Rule 'dist_or_over_and2' applied. ID: 30: ((((a ∨ c) ∧ (b ∨ ¬a)) ∨ d) ∧ ((b ∨ c) ∨ d)) was ((((a ∨ c) ∧ (b ∨ ¬a)) ∧ (b ∨ c)) ∨ d) at 23
Rule 'tautology' applied. ID: 31: true was (a ∨ ¬a) at 14
Rule 'dist_or_over_and2' applied. ID: 34: (((a ∨ c) ∨ a) ∧ ((b ∨ ¬a) ∨ a)) was (((a ∨ c) ∧ (b ∨ ¬a)) ∨ a) at 25
Rule 'dist_or_over_and2' applied. ID: 37: (((a ∨ c) ∨ d) ∧ ((b ∨ ¬a) ∨ d)) was (((a ∨ c) ∧ (b ∨ ¬a)) ∨ d) at 28
CNF: (((((((a ∨ c) ∨ a) ∧ ((a ∨ c) ∨ d)) ∧ ((b ∨ c) ∨ a)) ∧ ((b ∨ c) ∨ d)) ∧ ((b ∨ ¬a) ∨ a)) ∧ ((b ∨ ¬a) ∨ d))
BDD simplified: ((a ∧ (b ∨ d)) ∨ (¬a ∧ c))
Rule 'special_cnf_case' applied. ID: 21: ((((a ∨ ¬a) ∧ (a ∨ c)) ∧ (b ∨ ¬a)) ∧ (b ∨ c)) was ((a ∧ b) ∨ (¬a ∧ c)) at 8
Rule 'cnf_tautology_elimination' applied. ID: 15: (a ∨ c) was ((a ∨ ¬a) ∧ (a ∨ c)) at 16
Rule 'dist_or_over_and' applied. ID: 24: (((((a ∨ c) ∧ (b ∨ ¬a)) ∧ (b ∨ c)) ∨ a) ∧ ((((a ∨ c) ∧ (b ∨ ¬a)) ∧ (b ∨ c)) ∨ d)) was ((((a ∨ c) ∧ (b ∨ ¬a)) ∧ (b ∨ c)) ∨ (a ∧ d)) at 12
Rule 'dist_or_over_and2' applied. ID: 27: ((((a ∨ c) ∧ (b ∨ ¬a)) ∨ a) ∧ ((b ∨ c) ∨ a)) was ((((a ∨ c) ∧ (b ∨ ¬a)) ∧ (b ∨ c)) ∨ a) at 22
Rule 'dist_or_over_and2' applied. ID: 30: ((((a ∨ c) ∧ (b ∨ ¬a)) ∨ d) ∧ ((b ∨ c) ∨ d)) was ((((a ∨ c) ∧ (b ∨ ¬a)) ∧ (b ∨ c)) ∨ d) at 23
Rule 'tautology' applied. ID: 31: true was (a ∨ ¬a) at 14
Rule 'dist_or_over_and2' applied. ID: 34: (((a ∨ c) ∨ a) ∧ ((b ∨ ¬a) ∨ a)) was (((a ∨ c) ∧ (b ∨ ¬a)) ∨ a) at 25
Rule 'dist_or_over_and2' applied. ID: 37: (((a ∨ c) ∨ d) ∧ ((b ∨ ¬a) ∨ d)) was (((a ∨ c) ∧ (b ∨ ¬a)) ∨ d) at 28
Equivalent to original: Yes
Result: "Done"
